### ISSUE TITLE
Correct PixelFormat dependent errors

### ISSUE THREAD

**Issue by alexarchen** [DjvuNet/DjvuNet#129]
There are some Pixel format caused bugs in DjvuNet that leads to image scew and AccessViolationException. 

in class  DjvuNet.Graphics.Map:

```C#
    public static unsafe System.Drawing.Bitmap CopyDataToBitmap(int width, int height, IntPtr data, long length, PixelFormat format)
    {

            try
            {
                bmp = new System.Drawing.Bitmap(width, height, format);
                bmpData = bmp.LockBits(new System.Drawing.Rectangle(0, 0, bmp.Width, bmp.Height),
                                     ImageLockMode.WriteOnly, bmp.PixelFormat);

                MemoryUtilities.MoveMemory(bmpData.Scan0, data, length);
            }
```
This is incorrect because bmpData.dataStride must be taken into account when bmp.FixelFormat is 8 or 24 bit. In Map there is 1-byte data alignement, however Bitmap use 4-byte data alignment.
My solution is:
```C#
          try
          {
                bmp = new System.Drawing.Bitmap(width, height, format);
                bmpData = bmp.LockBits(new System.Drawing.Rectangle(0, 0, bmp.Width, bmp.Height),
                                     ImageLockMode.WriteOnly, bmp.PixelFormat);

                int datastride = (int)(length / height); // number of bytes in data row
                int bmpstride = bmpData.Stride; // number of dwords in bmp row
                for (int q=0;q<height;q++)
                 MemoryUtilities.MoveMemory((void *)(bmpData.Scan0+bmpstride*q), ((byte *)data)+datastride*q, datastride);
          }

```

**Comment by alexarchen** [comment]
In class DjvuNet.DjvuImage:

```
public unsafe System.Drawing.Bitmap BuildImage(int subsample = 1)
{
```
 you lock bitmap data with their native pixel format
 ```
                       BitmapData backgroundData =
                            background.LockBits(new System.Drawing.Rectangle(0, 0, background.Width, background.Height),
                                                ImageLockMode.ReadWrite, background.PixelFormat);
...

                        BitmapData foregroundData =
                            foreground.LockBits(new System.Drawing.Rectangle(0, 0, foreground.Width, foreground.Height),
                                                ImageLockMode.ReadOnly, foreground.PixelFormat);
```
and then it is used as being 32bpp format, as uint is 32 bit size

```
                            uint* backgroundRow = (uint*)(backgroundData.Scan0 + (yb * backgroundData.Stride));
                            uint* foregroundRow = (uint*)(foregroundData.Scan0 + (yf * foregroundData.Stride));
```
You can correct it in two ways: change pixel format for LockBits or take it into account when calculating pointer offset.
I decided to select first way like this:
```
BitmapData backgroundData =
                            background.LockBits(new System.Drawing.Rectangle(0, 0, background.Width, background.Height),
                                                ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
                        BitmapData foregroundData =
                            foreground.LockBits(new System.Drawing.Rectangle(0, 0, foreground.Width, foreground.Height),
                                                ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);

```
because it also allows you not to take care about Color Palette if PixelFormat.Format8bppIndexed

**Comment by 4creators** [comment]
Thanks for reporting these problems. I have not analyzed your solution but I have the following questions:

1. Have you tried to run your fixed DjvuNet against existing test framework?
2. Have you tested that all test files are displayed correctly?

If yes would you be interested in preparing PR to fix existing problems?

**Comment by alexarchen** [comment]
No, I haven't jet. I'm not sure my solutions are optimal, they are just hints. I'm testing on pdf2djvu created files. By the way, I found that Foreground mask color palette is not realized in DjvuNet so my last solution about BuildImage method is not exact. Mask can be colored. I think there was an issue. I'll try to do it

**Comment by alexarchen** [comment]
Finally I propose following solution of this issue and partially #60 issue:

in DjvuImage it's better to split backgroud and foreground building because sometimes only one layer is needed. Also creating white or black bitmap is memory consuming,  it's better to return null is there is no layer. If we do this, BuildImage function becomes clear and short, it just combines two images using Graphics functions:

```
        public unsafe System.Drawing.Bitmap BuildImage(int subsample = 1)
        {
            Verify.SubsampleRange(subsample);
            Stopwatch stopWatch = Stopwatch.StartNew();
            Bitmap back = BuildBackgroundImage(subsample);
            stopWatch.Stop();
            stopWatch.Restart();
            Bitmap fore = BuildForegroundImage(subsample);
            if (back == null) return fore;
            if (fore == null) return back;
            stopWatch.Stop();
            stopWatch.Restart();

            // combining
            Bitmap res = new Bitmap(_Page.Width / subsample, _Page.Height / subsample, PixelFormat.Format32bppArgb);

            using (System.Drawing.Graphics g = System.Drawing.Graphics.FromImage((Image)res))
            {
                g.InterpolationMode = InterpolationMode.Bicubic;
                g.DrawImage(_IsInverted ? InvertImage(back) : back, 0, 0, res.Width, res.Height);
                g.DrawImage(fore, 0, 0, res.Width, res.Height);
            }

            stopWatch.Stop();
            return res;
        }
```
So we need to add two additional functions,
simple for background:

```
public unsafe System.Drawing.Bitmap BuildBackgroundImage(int subsample = 1)
        {
            Verify.SubsampleRange(subsample);

            lock (_LoadingLock)
            {
                BG44Chunk[] backgrounds = _Page.PageForm?.GetChildrenItems<BG44Chunk>();
                if ((backgrounds == null) || (backgrounds.Length == 0)) return null;

                Stopwatch stopWatch = Stopwatch.StartNew();

                System.Drawing.Bitmap background = GetBackgroundImage(subsample, false);

                stopWatch.Stop();

                if (_IsInverted)
                    return InvertImage(background);
                return background;
            }
        }

```
And more complicated for foreground:

```
        public unsafe System.Drawing.Bitmap BuildForegroundImage(int subsample = 1)
        {
            Verify.SubsampleRange(subsample);

            
           lock (_LoadingLock)
            {
                if ((_Page.ForegroundIWPixelMap == null) && (_Page.ForegroundJB2Image == null)) return null;

                Stopwatch stopWatch = Stopwatch.StartNew();

                // TODO  if _Page.ForegroundIWPixelMap == null use only GetTextImage
                System.Drawing.Bitmap foreground=null;
                if (_Page.ForegroundIWPixelMap != null)
                    foreground = _Page.ForegroundIWPixelMap.GetPixelMap().ToImage();
                else return _Page.ForegroundJB2Image.GetColoredBitmap(_Page.ForegroundPalette);

                //using ( System.Drawing.Bitmap foreground = GetForegroundImage(subsample, false))
                using (foreground)
                {
                    stopWatch.Stop();
                    // TODO ETW logging goes here

                    stopWatch.Restart();

                    using (System.Drawing.Bitmap mask = _Page.ForegroundJB2Image.GetColoredBitmap(_Page.ForegroundPalette)) //GetTextImage(subsample, false))
                    {
                        stopWatch.Stop();
                        // TODO ETW logging goes here

                        

                        if (mask.PixelFormat != PixelFormat.Format8bppIndexed) throw new NotSupportedException("Mask Pixelformat must be 8bppIndexed");

                        stopWatch.Restart();

                        _HasLoaded = true;

                        Bitmap res = new Bitmap(mask.Width, mask.Height, PixelFormat.Format32bppArgb);

                            BitmapData resData =
                            res.LockBits(new System.Drawing.Rectangle(0, 0, mask.Width, mask.Height),
                             ImageLockMode.ReadWrite, res.PixelFormat);

                        BitmapData foregroundData = null;

                        if (foreground != null)
                            foregroundData = foreground.LockBits(new System.Drawing.Rectangle(0, 0, foreground.Width, foreground.Height),
                                                    ImageLockMode.ReadOnly, res.PixelFormat);

                        //int foregroundPixelSize = DjvuImage.GetPixelSize(foregroundData.PixelFormat);

                        BitmapData maskData = mask.LockBits(new System.Drawing.Rectangle(0, 0, mask.Width, mask.Height),
                                                            ImageLockMode.ReadOnly, mask.PixelFormat);

                        //int maskPixelSize = GetPixelSize(maskData);

                        int fgndHeight = foreground != null ? foreground.Height : mask.Height;
                        int fgndWidth = foreground != null ? foreground.Width : mask.Width;

                        int maskHeight = mask.Height;
                        int maskWidth = mask.Width;

                        int maskfgnH = maskHeight / fgndHeight;

                        int maskfgnW = maskWidth / fgndWidth;


                        for (int y = 0, yf = 0; y < maskHeight && yf < fgndHeight; y++)
                        {
                            byte* maskRow = (byte*)maskData.Scan0 + (y * maskData.Stride);
                            uint* foregroundRow = foreground != null ? (uint*)(foregroundData.Scan0 + (yf * foregroundData.Stride)) : null;
                            uint* resRow = (uint*)(resData.Scan0 + (yf * resData.Stride));


                            for (int x = 0, xf = 0; xf < fgndWidth; x++)
                            {
                                // Check if the mask byte is set
                                if (maskRow[x] != 0)
                                {
                                    uint c = 0xff000000;

                                    if (foreground != null)
                                        c = foregroundRow[xf];

                                    if (_IsInverted)
                                        c = InvertColor(c);

                                    resRow[xf] = c;
                                }
                                else
                                    resRow[xf] = maskRow[x];

                                if (x > 0)
                                {
                                    if (x % maskfgnW == 0)
                                        xf++;
                                }
                            }

                            if (y > 0)
                            {
                                if (y % maskfgnH == 0)
                                    yf++;
                            }
                        }
                        //});

                        mask.UnlockBits(maskData);
                        res.UnlockBits(resData);
                        if (foreground != null)
                            foreground.UnlockBits(foregroundData);

                        stopWatch.Stop();
                        // TODO ETW logging goes here

                        return res;
                    }
                }
            }
        } 

```
Take into account that there is  if (mask.PixelFormat != PixelFormat.Format8bppIndexed) throw new NotSupportedException("Mask Pixelformat must be 8bppIndexed") string because one needs to implement for 24bpp format. However this is rare case

Also there is one new function: JB2Image.GetColoredBitmap that should be implemented. I describe it in  issue #60.

**Comment by 4creators** [comment]
After analyzing the proposed solution I have arrived at the following conclusions:

1. The errors we had are due to wrong handling of padding as @alexarchen suggested and essentially it was the same error as in issue #57 - this already has been fixed with  PR #134
2. The remaining errors are due to lack of implementation of complex foreground composition images what is tracked by issue #60.

Due to the fact that proposed solution is not consistent with Djvu format specification, we cannot accept the proposed fix (anyway it's not PR) and we are closing this issue in favor of #60 which will track implementation of missing functionality.
